### PR TITLE
Abiv2 testing framework that uses Evmone

### LINKED ISSUE #6795: Create evmone runner
## Abstract

Create a wrapper around evmone for executing arbitrary code.

## Motivation

Needed for fuzzing abiv2 as an alternative for Geth.

### PR DESCRIPTION
**Pull request by bshastry**
### Description

This is an incomplete draft PR that should eventually close #6795 

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

### CODE REVIEW

File: test/tools/ossfuzz/abiV2Fuzzer.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	// If we cannot compile contract, we simply bail.
+	try
+	{
+		// Compile contract generated by the proto fuzzer
+		SolidityExecutionFramework executionFramework;
+		byteCode = executionFramework.compileContract(input);
+		Json::Value methodIdentifiers = executionFramework.getMethodIdentifiers();
+
+		// Call the first function
+	    for (auto const& id : methodIdentifiers.getMemberNames())
```
Comment (chfast) [inline test/tools/ossfuzz/abiV2Fuzzer.cpp:48]: Why do you run it in a loop if you are overwriting the `m`?

File: test/tools/ossfuzz/abiV2Fuzzer.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	    {
+	    	auto msg_input = EvmOneVM::from_hex(methodIdentifiers[id].asString());
```
Comment (chfast) [inline test/tools/ossfuzz/abiV2Fuzzer.cpp]: The `msg_input` will go out of scope.

File: test/tools/ossfuzz/abiV2Fuzzer.cpp (hunk 1 above)
Comment (bshastry) [reply test/tools/ossfuzz/abiV2Fuzzer.cpp:48]: The for loop is executed only once due to the break statement in the end. The idea is to call the first available function.

File: test/tools/ossfuzz/abiV2Fuzzer.cpp (hunk 2 above)
Comment (bshastry) [reply test/tools/ossfuzz/abiV2Fuzzer.cpp]: Ah, stupid mistake, thanks for pointing out. However, even after fixing this error, the revert still happens. So my code has more bugs!

File: test/tools/ossfuzz/abiV2FuzzerCommon.h [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+#include <liblangutil/SourceReferenceFormatter.h>
+#include <libdevcore/Keccak256.h>
+
+namespace dev
+{
+namespace test
+{
+namespace abiv2fuzzer
+{
+class SolidityExecutionFramework
```
Comment (chriseth) [inline test/tools/ossfuzz/abiV2FuzzerCommon.h:16]: Is it possible to extend the existing framework?

File: test/tools/ossfuzz/protoToAbiV2.cpp [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+}
+
+void ProtoConverter::visit(UnsignedIntegerValue const& _x)
+{
+	auto width = 8 * ((_x.width() % 32) + 1);
+	u256 value = generateUnsigned(_x.value64(), _x.value128(), _x.value192(), _x.value256());
+	switch (width)
+	{
+	case 8:
+		m_output << u8(value);
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToAbiV2.cpp:128]: Why do you need to convert to the smaller type here?

File: test/tools/evmoneRunner.cpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	auto code = from_hex(_runtimeCode);
+	result = vm->execute(vm, this, rev, &msg, code.data(), code.size());
+	cout << result.status_code << endl;
+	assert(result.status_code == EVMC_SUCCESS);
+	std::string output = std::string{reinterpret_cast<const char*>(result.output_data), result.output_size};
+	assert(output == "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01");
+	cout << result.output_data << endl;
+}
+
+std::basic_string<uint8_t> EvmOneVM::from_hex(std::string _hex)
```
Comment (chriseth) [inline test/tools/evmoneRunner.cpp:77]: We should have plenty of such conversion functions already.

File: test/tools/ossfuzz/abiV2Fuzzer.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+#include <fstream>
+#include <test/tools/evmoneRunner.h>
+#include "abiV2FuzzerCommon.h"
+
+using namespace dev::test::abiv2fuzzer;
+using namespace external::evmone;
+using namespace std;
+using namespace dev;
+
+extern "C" int LLVMFuzzerTestOneInput(uint8_t const* _data, size_t _size)
```
Comment (chriseth) [inline test/tools/ossfuzz/abiV2Fuzzer.cpp:27]: Is this the main entry function? If yes, why does it not use the ProtoConverter (or did I overlook something)?

**Comment by bshastry** [comment]
@chriseth I'm afraid the proto stuff is not ready for review yet. This PR was mainly meant for reviewing the evmone wrapper which itself is also not reviewable yet.

So, here's one proposal. Create separate (more easily reviewable) PRs for
  - [PR ready for review] evmone build CI integration (see #6803 )
  - [Upcoming PR]  evmone wrapper
  - [Upcoming PR] solidity wrapper (this accepts fuzzer input i.e., solidity and returns runtime bytecode)
  - [Upcoming PR] non-proto fuzzer harness to check if build works as intended
  - [Upcoming PR] proto fuzzer

Each of these PRs are meant to be incremental. Hope that's okay.

**Comment by chriseth** [comment]
Ok, sure!

**Comment by bshastry** [comment]
Closing this PR since it will be broken down.

**Pull request closed without merging**